rbac/management/group/relation_api_dual_write_group_handler.py

```diff
@@ -123,54 +129,59 @@ def _replicate(self):
+            V2role.for_system_role(str(system_role.uuid)),
+            V2boundresource(("rbac", "workspace"), str(self.default_workspace.id)),
+            groups=frozenset([str(self.group.uuid)]),
+        )
+        mapping = BindingMapping.for_role_binding(binding, system_role)
+        self.group_relations_to_add.extend(mapping.as_tuples())
+        return mapping
+
+    def generate_relations_to_add_roles(
+        self, roles: Iterable[Role], remove_default_access_from: Optional[TenantMapping] = None
+    ):
+        """Generate relations to add roles."""
         if not self.replication_enabled():
             return
 
         def add_group_to_binding(mapping: BindingMapping):
             self.group_relations_to_add.append(mapping.add_group_to_bindings(str(self.group.uuid)))
 
-        def create_default_mapping():
-            assert role.system is True, "Expected system role. Mappings for custom roles must already be created."
-            binding = V2rolebinding(
-                str(uuid4()),
-                # Assumes same role UUID for V2 system role equivalent.
-                V2role.for_system_role(str(role.uuid)),
-                V2boundresource(("rbac", "workspace"), str(self.default_workspace.id)),
-                groups=frozenset([str(self.group.uuid)]),
-            )
-            mapping = BindingMapping.for_role_binding(binding, role)
-            self.group_relations_to_add.extend(mapping.as_tuples())
-            return mapping
+        for role in roles:
+            self._update_mapping_for_role(role, update_mapping=add_group_to_binding)
+
+        if remove_default_access_from is not None:
+            default_binding = self._default_binding(mapping=remove_default_access_from)
+            self.group_relations_to_remove.append(default_binding)
+
+    def replicate(self):
+        """Replicate added role."""
+        if not self.replication_enabled():
+            return
 
-        self._update_mapping_for_role(
-            role, update_mapping=add_group_to_binding, create_default_mapping_for_system_role=create_default_mapping
-        )
         self._replicate()
 
-    def replicate_removed_role(self, role: Role):
+    def generate_relations_to_remove_roles(self, roles: Iterable[Role]):
         """Replicate removed role."""
         if not self.replication_enabled():
             return
 
-        self._update_mapping_for_role_removal(role)
-        self._replicate()
+        for role in roles:
+            self._update_mapping_for_role_removal(role)
 
     def _update_mapping_for_role_removal(self, role: Role):
         def remove_group_from_binding(mapping: BindingMapping):
             self.group_relations_to_remove.append(mapping.remove_group_from_bindings(str(self.group.uuid)))
 
-        self._update_mapping_for_role(
-            role, update_mapping=remove_group_from_binding, create_default_mapping_for_system_role=lambda: None
-        )
```

Comment:
Changing how this works I think may introduce a bug – if this code was called to remove a system role from a group where it wasn't assigned, it would end up getting added. I would keep the old logic where this could be controlled by the arguments to the method.

Reply:
I see. I removed last commit. I added comment.